src/lib/authentication/components/protectedRoute.tsx

```diff
@@ -0,0 +1,36 @@
+import { useNavigate, Route } from "react-router-dom";
+import { useAuthContext } from "./useAuthContext";
+
+type ProtectedRouteProps = {
+  path: string;
+  element: React.ReactElement;
+  permissions: string[];
+  fallbackPath?: string;
+};
+
+const ProtectedRoute = ({
+  path,
+  element,
+  permissions,
+  fallbackPath = "/",
+}: ProtectedRouteProps) => {
+  const navigate = useNavigate();
+  const { context } = useAuthContext();
+
+  // Check if the user has all the required permissions
+  const hasAllPermissions = permissions.every((permission) =>
+    context.hasPermission(permission)
+  );
+
+  if (!hasAllPermissions) {
+    navigate(fallbackPath);
```

Comment:
if, if the context change, indipendently if the browser is on this route, we trigger a navigate. Also, side effects in FuncionalComponents shouldn't be in a useEffect() ?